docs/migrations.rst

```diff
@@ -205,6 +205,84 @@ Queries can be executed with the ``execute()`` and ``query()`` methods. The
     the ``execute()`` command. This is especially important when using
     DELIMITERs during insertion of stored procedures or triggers which
     don't support DELIMITERs.
+
+Parameterised queries
+~~~~~~~~~~~~~~~~~~~~~
+
+Both the ``execute()`` and ``query()`` methods accept a second argument for an array of values for parameterised queries.
+
+The following example shows queries with positional ? placeholders.
+
+.. code-block:: php
+
+        <?php
+
+        use Phinx\Migration\AbstractMigration;
+
+        class MyNewMigration extends AbstractMigration
+        {
+            /**
+             * Migrate Up.
+             */
+            public function up()
+            {
+                $queryParams = array(10, 15);
+
+                // execute()
+                $count = $this->execute('DELETE FROM users WHERE user_id BETWEEN ? AND ?', $queryParams);
+
+                // query()
+                $rows = $this->query('SELECT * FROM users WHERE user_id BETWEEN ? AND ?', $queryParams);
+            }
+
+            /**
+             * Migrate Down.
+             */
+            public function down()
+            {
+
+            }
+        }
+
+This example show queries with named placeholders and an optional bind type.
+
+.. code-block:: php
+
+        <?php
+
+        use Phinx\Migration\AbstractMigration;
+
+        class MyNewMigration extends AbstractMigration
+        {
+            /**
+             * Migrate Up.
+             */
+            public function up()
+            {
+                $queryParams = array(
```

Comment:
So this is the same query format that I'd recommended in the RFC way back, so I clearly think this is the _good_ solution, but I worry about arrays, especially multi-dimensional arrays, as a way of specifying behaviour. Phinx is already doing a tonne of this (such as [setting column attributes in a column definition](http://docs.phinx.org/en/latest/migrations.html#creating-a-table)). My problem is that the user has to know this particular data structure to define the correct behaviour, in this case bind-typing. This is some dense logic:

> If a `$queryParams` item value is not an array, then the array item value is the bind-value. Otherwise, the item-value array must have an item with key `value` to specify the bind-value, and optionally an item with key `type` specifying the bind-type.

Gross. Try getting your dev team to internalise that logic. And I learn this where, in the documentation, or by reading examples? Not to mention, if this migration tool starts supporting [JSON column types](https://mariadb.com/kb/en/mariadb/column_json/), the bind value might legitimately be an array….

So what about an object oriented approach, where the binds are all defined in classes? The method signatures explain the behaviour, and where there's any confusion, the docblocks explain the rest:

```php
// This example has sorely insufficient documentation
interface QueryBind {
    const TYPE_INT = \PDO::PARAM_INT;
    const TYPE_BOOL = \PDO::PARAM_BOOL; // etc…
    /**
     * @param mixed $bindValue
     */
    public function __construct($bindValue);
    /**
     * @param int $bindType Should match QueryBind::TYPE_*
     * @return $this
     * @see \PDO
     * @link http://php.net/manual/en/pdo.constants.php
     */
    public function setBindType($bindType);
    /**
     * @return int By default `static::TYPE_STR`
     * @link http://php.net/manual/en/pdo.constants.php
     */
    public function getBindType();
}

class AbstractMigration {
    /**
     * @param string $query
     * @param QueryBind[] $binds
     */
    public function execute($query, array $binds = null){ /* magic */ }
}

class SomeMigration extends AbstractMigration{
    public function up(){
        $binds = [
            'bindA' => new QueryBind('somestring'), // default is string bind
            'bindB' => (new QueryBind(5))->setBindType(QueryBind::TYPE_INT),
            'bindC' => new QueryBind(5, QueryBind::TYPE_INT), // or perhaps this as an alternative?
        ];
        $this->execute('SELECT … WHERE a=:bindA AND b=:bindB AND c=:bindC', $binds);
    }
}
```

The benefit here is the behaviour is prescribed by the object's methods and parameters, rather than in some array format.

Anyway, all this to say that when I did my version I still went with the array model. I just thought OOO alternative version should be discussed.